File: proxy.py
#!/usr/bin/env python3
"""
GitCode CORS Proxy for CANN Recipes Blog
Zero dependencies — uses only Python stdlib.

Usage:
    python proxy.py

Then open index.html in your browser.
"""
import http.server
import urllib.request
import urllib.parse
import json
import base64
import sys

PORT = 8081

class ProxyHandler(http.server.BaseHTTPRequestHandler):
    def _cors_headers(self):
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', '*')

    def do_OPTIONS(self):
        self.send_response(204)
        self._cors_headers()
        self.end_headers()

    def do_GET(self):
        params = dict(urllib.parse.parse_qsl(
            urllib.parse.urlparse(self.path).query
        ))
        repo   = params.get('repo', '')
        path   = params.get('path', '')
        branch = params.get('branch', 'master')

        if not repo or not path:
            self.send_response(400)
            self._cors_headers()
            self.send_header('Content-Type', 'text/plain')
            self.end_headers()
            self.wfile.write(b'Missing repo or path param.\n'
                b'Example: /?repo=cann-recipes-infer'
                b'&path=docs/models/deepseek-r1/deepseek_r1_decode_optimization.md'
                b'&branch=master')
            return

        ns = 'cann'
        repo_id = f'{ns}%2F{repo}'
        api_url = (
            f'https://web-api.gitcode.com/api/v2/projects/{repo_id}/repository/files'
            f'?repoId={urllib.parse.quote(repo_id)}'
            f'&ref={urllib.parse.quote(branch)}'
            f'&file_path={urllib.parse.quote(path)}'
            f'&ref_replace_web={urllib.parse.quote(branch)}'
        )

        req = urllib.request.Request(api_url, headers={
            'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) '
                          'AppleWebKit/537.36 (KHTML, like Gecko) '
                          'Chrome/120.0.0.0 Safari/537.36',
            'Accept': 'application/json, text/plain, */*',
            'Origin': 'https://gitcode.com',
            'Referer': f'https://gitcode.com/{ns}/{repo}/blob/{branch}/{path}',
        })

        try:
            with urllib.request.urlopen(req, timeout=30) as resp:
                data = json.loads(resp.read().decode('utf-8'))

            if 'content' not in data:
                raise ValueError('No content field in API response')

            body = base64.b64decode(data['content'])
            content_type = data.get('file_type', 'application/octet-stream')

            self.send_response(200)
            self._cors_headers()
            self.send_header('Content-Type', content_type)
            self.send_header('Content-Length', str(len(body)))
            self.send_header('Cache-Control', 'public, max-age=300')
            self.end_headers()
            self.wfile.write(body)

        except Exception as e:
            msg = f'Fetch failed: {e}'.encode('utf-8')
            print(f'[proxy] error for {repo}/{path}: {e}', file=sys.stderr)
            self.send_response(502)
            self._cors_headers()
            self.send_header('Content-Type', 'text/plain')
            self.end_headers()
            self.wfile.write(msg)

    def log_message(self, fmt, *args):
        print(f'[proxy] {args[0]}')


if __name__ == '__main__':
    server = http.server.HTTPServer(('127.0.0.1', PORT), ProxyHandler)
    print(f'\n  GitCode Proxy running at http://localhost:{PORT}')
    print(f'  Open index.html in your browser.\n')
    try:
        server.serve_forever()
    except KeyboardInterrupt:
        print('\nStopped.')

File: proxy/proxy.js
#!/usr/bin/env node
/**
 * GitCode Proxy Server for CANN Recipes Blog
 *
 * Why needed:
 *   web-api.gitcode.com sets Access-Control-Allow-Origin: https://gitcode.com,
 *   so a browser on any other origin (file://, localhost, etc.) is blocked.
 *   This proxy forwards the request server-side and re-serves the file content
 *   with permissive CORS headers.
 *
 * How to start:
 *   node proxy.js
 *
 * Then open index.html in any browser.
 */

const http  = require('http');
const https = require('https');

const PORT = 8081;

function fetchFile(repo, filePath, branch) {
  return new Promise((resolve, reject) => {
    const ns      = 'cann';
    const repoId  = `${ns}%2F${repo}`;
    const apiUrl  =
      `https://web-api.gitcode.com/api/v2/projects/${repoId}/repository/files` +
      `?repoId=${encodeURIComponent(repoId)}` +
      `&ref=${encodeURIComponent(branch)}` +
      `&file_path=${encodeURIComponent(filePath)}` +
      `&ref_replace_web=${encodeURIComponent(branch)}`;

    const req = https.request(apiUrl, {
      headers: {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36',
        'Accept': 'application/json, text/plain, */*',
        'Origin':   'https://gitcode.com',
        'Referer':  `https://gitcode.com/${ns}/${repo}/blob/${branch}/${filePath}`,
      },
    }, res => {
      const chunks = [];
      res.on('data', c => chunks.push(c));
      res.on('end', () => {
        try {
          const json = JSON.parse(Buffer.concat(chunks).toString('utf-8'));
          if (!json.content) {
            return reject(new Error('No content field – response: ' +
              JSON.stringify(json).substring(0, 200)));
          }
          const buf  = Buffer.from(json.content, 'base64');
          const type = json.file_type || 'application/octet-stream';
          resolve({ buf, type });
        } catch (e) {
          reject(e);
        }
      });
    });
    req.on('error', reject);
    req.end();
  });
}

function fetchTree(repo, filePath, branch) {
  return new Promise((resolve, reject) => {
    const ns      = 'cann';
    const repoId  = `${ns}%2F${repo}`;
    const apiUrl  =
      `https://web-api.gitcode.com/api/v2/projects/${repoId}/repository/upper_files_tree` +
      `?repoId=${encodeURIComponent(repoId)}` +
      `&ref_name=${encodeURIComponent(branch)}` +
      `&file_path=${encodeURIComponent(filePath)}`;

    const req = https.request(apiUrl, {
      headers: {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36',
        'Accept': 'application/json, text/plain, */*',
        'Origin':   'https://gitcode.com',
        'Referer':  `https://gitcode.com/${ns}/${repo}/tree/${branch}/${filePath}`,
      },
    }, res => {
      const chunks = [];
      res.on('data', c => chunks.push(c));
      res.on('end', () => {
        try {
          const json = JSON.parse(Buffer.concat(chunks).toString('utf-8'));
          // Normalize to [{name, type}] — handle array or wrapped response
          const raw = Array.isArray(json) ? json : (json.data || json.trees || json.items || []);
          const items = raw
            .map(item => ({
              name: item.name || item.fileName || item.file_name || '',
              type: item.type === 'blob' ? 'blob'
                  : item.type === 'tree' ? 'tree'
                  : (item.name || '').includes('.') ? 'blob' : 'tree',
            }))
            .filter(item => item.name);
          resolve(items);
        } catch (e) {
          reject(new Error('Tree parse error: ' + e.message));
        }
      });
    });
    req.on('error', reject);
    req.end();
  });
}

http.createServer(async (req, res) => {
  // Allow any local origin (file://, localhost, etc.)
  res.setHeader('Access-Control-Allow-Origin',  '*');
  res.setHeader('Access-Control-Allow-Methods', 'GET, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', '*');

  if (req.method === 'OPTIONS') {
    res.writeHead(204);
    res.end();
    return;
  }

  if (req.method !== 'GET') {
    res.writeHead(405, { 'Content-Type': 'text/plain' });
    res.end('Method Not Allowed');
    return;
  }

  const q = Object.fromEntries(new URL(req.url, 'http://localhost').searchParams);

  if (!q.repo || !q.path) {
    res.writeHead(400, { 'Content-Type': 'text/plain' });
    res.end('Missing required query params: repo, path\n' +
            'Example: /?repo=cann-recipes-infer&path=docs/models/deepseek-r1/deepseek_r1_decode_optimization.md&branch=master');
    return;
  }

  if (q.action === 'tree') {
    try {
      const items = await fetchTree(q.repo, q.path, q.branch || 'master');
      const json  = JSON.stringify(items);
      res.writeHead(200, {
        'Content-Type':   'application/json',
        'Content-Length': Buffer.byteLength(json),
      });
      res.end(json);
    } catch (err) {
      console.error('[proxy] tree error:', err.message);
      res.writeHead(502, { 'Content-Type': 'text/plain' });
      res.end('Tree fetch failed: ' + err.message);
    }
    return;
  }

  try {
    const { buf, type } = await fetchFile(q.repo, q.path, q.branch || 'master');
    res.writeHead(200, {
      'Content-Type':   type,
      'Content-Length': buf.length,
    });
    res.end(buf);
  } catch (err) {
    console.error('[proxy] error:', err.message);
    res.writeHead(502, { 'Content-Type': 'text/plain' });
    res.end('Upstream fetch failed: ' + err.message);
  }

}).listen(PORT, '127.0.0.1', () => {
  console.log(`\nGitCode Proxy running at http://localhost:${PORT}`);
  console.log('Open index.html in your browser.\n');
});
